### ISSUE TITLE
Assembling a pipeline from pre-fitted components

### ISSUE THREAD

**Issue by jhaneyrf** [jpmml/sklearn2pmml#391]
I'm trying to take an XGBoost model that I've built and create a PMML file. To save work for other members on my team, I would like the PMML file to handle the data preprocessing.

For this model, I did no missing value imputation, but I had two categorical variables that I applied a custom mapping to.

I was hoping that this code would work, but I'm getting an error. Do you have any suggestions for how I can handle this?

I'm on sklearn2pmml version 0.97 and scikit-learn version 1.2.2.

The code (anonymized):
```python
import pickle

import pandas as pd

from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn2pmml import PMMLPipeline, sklearn2pmml
from sklearn2pmml.preprocessing import LookupTransformer

MODEL_VARIABLES_DE = [f"variable{val}" for val in range(1, 132)]

STATE_DICT = {
    "KY": 0,
    "GA": 1,
    "LA": 2,
    "MS": 3,
    "SC": 4,
    "TN": 5,
    "AL": 6,
    "TX": 7}


VARIABLE22_DICT = {
        "A": 0,
        "B": 1,
        "C": 2,
        "D": 3}


PREPROCESSOR = ColumnTransformer(
    transformers=[
        ('state_encoder', LookupTransformer(STATE_DICT, default_value=7),
         ["variable21"]),
        ('customer_type_encoder', LookupTransformer(
            VARIABLE22_DICT, default_value=3), ["variable22"])],
    remainder="passthrough")

PREPROCESSOR.fit(pd.DataFrame({
    "variable21": ["AL", "GA", "KY", "LA", "MS", "SC", "TN", "TX"],
    "variable22": ["NEW", "CONV", "RENEW", "FC"] * 2}))

def load_model(filename=None):
    """Load the model pickle file"""
    if filename is None:
        filename = "model.pkl"
    with open(filename, "rb") as pickle_file:
        model = pickle.load(pickle_file)
    return model

MODEL = load_model()

PMML_PIPELINE = PMMLPipeline(steps=[
    ('preprocessor', PREPROCESSOR), ('classifier', MODEL)])
sklearn2pmml(PMML_PIPELINE, "model.pmml")
```

The error:
```
Standard error:
Exception in thread "main" java.lang.IllegalArgumentException: The transformer object (Java class java.util.ArrayList) is not a supported Transformer
        at org.jpmml.python.CastFunction.apply(CastFunction.java:47)
        at sklearn.pipeline.Pipeline$1.apply(Pipeline.java:106)
        at sklearn.pipeline.Pipeline$1.apply(Pipeline.java:97)
        at com.google.common.collect.Lists$TransformingRandomAccessList$1.transform(Lists.java:631)
        at com.google.common.collect.TransformedIterator.next(TransformedIterator.java:52)
        at sklearn.Composite.encodeFeatures(Composite.java:138)
        at sklearn2pmml.pipeline.PMMLPipeline.encodePMML(PMMLPipeline.java:154)
        at com.sklearn2pmml.Main.run(Main.java:82)
        at com.sklearn2pmml.Main.main(Main.java:65)
Caused by: java.lang.ClassCastException: Cannot cast java.util.ArrayList to sklearn.Transformer
        at java.lang.Class.cast(Unknown Source)
        at org.jpmml.python.CastFunction.apply(CastFunction.java:45)
        ... 8 more
```

**Comment by jhaneyrf** [comment]
Sorry for the formatting, I didn't think about what Markdown would do to my terminal output.

**Comment by vruusmann** [comment]
TLDR: The culprit is the `model = pickle.load(pickle_file)` line, because it returns some array-like object instead of a Scikit-Learn estimator object.

The Sklearn2PMML/JPMML-SkLearn side of the software stack is OK. There is something wrong with the objects that you've put inside your `PMML_PIPELINE` object.  If you tried to call `PMML_PIPELINE.predict(X)`, then this would probably fail too.

The feature transformation part is actually OK. If you assemble a transformer-only pipeline, then the conversion to PMML succeeds:

```python
PMML_PIPELINE = PMMLPipeline(steps=[
	('preprocessor', PREPROCESSOR)
])

sklearn2pmml(PMML_PIPELINE, "preprocessor.pmml")
```

Try to figure out what's wrong with your pickled model object:
```python
with open(filename, "rb") as pickle_file:
	model = pickle.load(pickle_file)
# THIS!
print(model)
```

I suspect there is some Pickle data format-level issue happening. Perhaps the model was pickled using `joblib.dump()` or `dill.dump()` method calls, and when you train to unpickle it using a plain `pickle.load()` method call, you only get a partial result back.

**Comment by jhaneyrf** [comment]
Thanks! I don't know how long it would have taken me to see this myself.
